### ISSUE TITLE
Incorrect Implementation of WriteMessage and ReadMessage

### ISSUE THREAD

**Issue by mimoo** [flynn/noise#19]
You are not following the spec in your [WriteMessage](https://github.com/flynn/noise/blob/master/state.go#L330) and [ReadMessage](https://github.com/flynn/noise/blob/master/state.go#L387) implementations.

For example, when reading a [ES](https://github.com/flynn/noise/blob/master/state.go#L432) token, you are suppose to choose what keys to use based on who is the initiator. Here is what you do:

```go
case MessagePatternDHES:
    s.ss.MixKey(s.ss.cs.DH(s.s.Private, s.re))
```

Here is what you should [according to the spec](http://noiseprotocol.org/noise.html#the-handshakestate-object):

> For "es": Calls MixKey(DH(e, rs)) if initiator, MixKey(DH(s, re)) if responder.

**Comment by mimoo** [comment]
MMMmm, I think I see what you're doing, you inverted the patterns so that they would work for your code depending on if you use Write or Read

**Comment by titanous** [comment]
Thanks for pointing this out. I did some spelunking and found that the reason for this is that it was originally specified this way: https://github.com/noiseprotocol/noise_spec/blob/e9b7a658258a583bafe1b0de790a13ef99a2a297/noise.md

The spec has since changed and I agree that it makes reviewing the code against the specification confusing. I won't have time to fix this immediately, so if you can submit a PR it'd be appreciated!

**Comment by titanous** [comment]
Fixed by #22.
